### PR TITLE
TE-3767: creating rule to check component name

### PR DESCRIPTION
**Pull request by KostKravchenko**
Ticket: https://spryker.atlassian.net/browse/TE-3767

### CODE REVIEW

File: src/collector/components/component.ts [hunk 1]
```diff
@@ -27,9 +28,43 @@ export interface IComponent<M extends IFile = IFile, W extends IFile = IFile, S
     }
 }
 
+function extractCorrectTwigFileName(defaultFileName: string, directoryFiles: string[]) {
+    const twigFileNameInComponent = directoryFiles.filter(fileName => fileName.indexOf('.twig') >= 0)[0];
```
Comment (supproduction) [inline src/collector/components/component.ts]: Suggest to replace fileName.indexOf('.twig') >= 0 -> fileName.includes('.twig')

File: src/collector/components/component.ts [hunk 2]
```diff
@@ -27,9 +28,43 @@ export interface IComponent<M extends IFile = IFile, W extends IFile = IFile, S
...
+    if (!directoryFiles.includes(defaultFileName) && twigFileNameInComponent) {
+        return twigFileNameInComponent;
+    }
+
+    return defaultFileName;
+}
+
+function extractCorrectScssFileName(defaultFileName: string, directoryFiles: string[]) {
+    const scssFileNameInComponent = directoryFiles.filter(fileName =>
+        fileName.indexOf('.scss') >= 0 && fileName !== 'style.scss'
```
Comment (supproduction) [inline src/collector/components/component.ts]: Suggest to replace fileName.indexOf('. scss') >= 0 -> fileName.includes('. scss')

File: src/collector/components/component.ts [hunk 3]
```diff
@@ -27,9 +28,43 @@ export interface IComponent<M extends IFile = IFile, W extends IFile = IFile, S
...
+    if (!directoryFiles.includes(defaultFileName) && scssFileNameInComponent) {
+        return scssFileNameInComponent;
+    }
+
+    return defaultFileName;
+}
+
+function extractCorrectScriptFileName(defaultFileName: string, directoryFiles: string[]) {
+    const scriptFileNameInComponent = directoryFiles.filter(fileName =>
+        fileName.indexOf('.ts') >= 0 && fileName !== 'index.ts'
```
Comment (supproduction) [inline src/collector/components/component.ts]: Suggest to replace fileName.indexOf('. ts') >= 0 -> fileName.includes('. ts')

**Review by supproduction — CHANGES_REQUESTED** [review]

**Review by supproduction — APPROVED** [review]

File: rules/correct-component-name.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+const { Rule } = require('../api');
+
+module.exports = class extends Rule {
+    getName() {
+        return 'correct-component-name';
+    }
+
+    checkTwigFile(component) {
+        const { twig } = component.files;
+        const {name: twigFileName} = twig;
```
Comment (tolerants) [inline rules/correct-component-name.js]: Please add spaces after `{` and before `}`.

File: rules/correct-component-name.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+        if (configName && configName !== name) {
+            this.outcome.addError(`There is wrong name property in config block of twig file in ${type} ${bold(name)}:\n${dim(path)}`);
+        }
+    }
+
+    checkScssFile(component) {
+        const { sass } = component.files;
+        const { mixins } = sass.api.external;
+        const {name: scssFileName} = sass;
```
Comment (tolerants) [inline rules/correct-component-name.js]: Please add spaces after `{` and before `}`.

File: rules/correct-component-name.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+    checkTypescriptFile(component) {
+        if (!component.files.typescript.api.external) {
+            return
+        };
+
+        const { typescript } = component.files;
+        const { external } = typescript.api;
+        const { classes } = external;
+        const {name: typescriptFileName} = typescript;
```
Comment (tolerants) [inline rules/correct-component-name.js]: Please add spaces after `{` and before `}`.

File: rules/correct-component-name.js [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+                configName = contractStringFromNameProperty.slice(0, contractStringFromNameProperty.indexOf('\''));
+            }
+        }
+
+        if (name !== twigFileName.slice(0, twigFileName.lastIndexOf('.twig'))) {
+            this.outcome.addError(`There is wrong name of twig file in ${type} ${bold(name)}:\n${dim(path)}`);
+        }
+
+        if (!configDefinition) {
+            this.outcome.addError(`There is no config block in twig file of ${type} ${bold(name)}:\n${dim(path)}`);
```
Comment (tolerants) [inline rules/correct-component-name.js]: I think this check is not related to component name, I think we need some basic checks, like `correct-twig-structure`, `correct-scss-structure` and `correct-ts-structure`, where we can check if component contains `config` and `data` objects, if main block name is `body`, etc.

File: rules/correct-component-name.js [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    getName() {
+        return 'correct-component-name';
+    }
+
+    checkTwigFile(component) {
+        const { twig } = component.files;
+        const {name: twigFileName} = twig;
+        const configDefinition = twig.api.external.definitions.filter(definition => definition.name === 'config')[0];
+        const { type, name, path } = component;
+        let contract;
```
Comment (tolerants) [inline rules/correct-component-name.js]: Why do you need this variable in this scope?

File: rules/correct-component-name.js [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+                this.checkTwigFile(component);
+            }
+
+            if (files.sass.exists) {
+                this.checkScssFile(component);
+            }
+
+            if (files.typescript.exists) {
+                this.checkTypescriptFile(component);
+            }
```
Comment (tolerants) [inline rules/correct-component-name.js]: Please add folder name check.

File: rules/correct-component-name.js [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+        const { sass } = component.files;
+        const { mixins } = sass.api.external;
+        const {name: scssFileName} = sass;
+        const { type, name, path } = component;
+
+        if (name !== scssFileName.slice(0, scssFileName.lastIndexOf('.scss'))) {
+            this.outcome.addError(`There is wrong name of scss file in ${type} ${bold(name)}:\n${dim(path)}`);
+        }
+
+        mixins.forEach(mixin => {
```
Comment (tolerants) [inline rules/correct-component-name.js]: This functionality is also not related to correct files name.

File: rules/correct-component-name.js [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+        const { external } = typescript.api;
+        const { classes } = external;
+        const {name: typescriptFileName} = typescript;
+        const { type, name, path } = component;
+
+        if (name !== typescriptFileName.slice(0, typescriptFileName.lastIndexOf('.ts'))) {
+            this.outcome.addError(`There is wrong name of typescript file in ${type} ${bold(name)}:\n${dim(path)}`);
+        }
+
+        classes.forEach(singleClass => {
```
Comment (tolerants) [inline rules/correct-component-name.js]: The same.

**Review by tolerants — CHANGES_REQUESTED** [review]

File: rules/correct-component-name.js [hunk 12]
```diff
@@ -0,0 +1,109 @@
+const { dim, bold } = require('colors');
+const { Rule } = require('../api');
+
+module.exports = class extends Rule {
+    getName() {
+        return 'correct-component-name';
+    }
+
+    checkTwigFile(component) {
+        const { twig } = component.files;
```
Comment (tolerants) [inline rules/correct-component-name.js]: Please check views and templates.

**Review by tolerants — CHANGES_REQUESTED** [review]

File: rules/correct-component-name.js (hunk 7 above)
Comment (KostKravchenko) [reply rules/correct-component-name.js]: I created separate rules for checking structure of component twig, scss and typescript files

File: rules/correct-component-name.js (hunk 8 above)
Comment (KostKravchenko) [reply rules/correct-component-name.js]: It is my bad.
I declared this variable in another scope

File: rules/correct-component-name.js (hunk 9 above)
Comment (KostKravchenko) [reply rules/correct-component-name.js]: I think, it will be better to check folder name in another rule.

File: rules/correct-component-name.js (hunk 10 above)
Comment (KostKravchenko) [reply rules/correct-component-name.js]: I created rule 'correct-scss-component-structure' for this logic.

File: rules/correct-component-name.js (hunk 11 above)
Comment (KostKravchenko) [reply rules/correct-component-name.js]: I created rule 'correct-typescript-component-structure' for this logic

File: rules/correct-component-name.js (hunk 12 above)
Comment (KostKravchenko) [reply rules/correct-component-name.js]: I think, it will be better check views and template in another rule

File: rules/correct-scss-component-structure.js [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+module.exports = class extends Rule {
+    getName() {
+        return 'correct-scss-component-structure';
+    }
+
+    test(data) {
+        parseOutputFieldHelper(data.components).forEach(component => {
+            const { module, files, type, name, path } = component;
+
+            if (files.sass.exists) {
```
Comment (FRITS-Kh) [inline rules/correct-scss-component-structure.js]: Please use early return.

File: rules/correct-twig-component-structure.js [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+module.exports = class extends Rule {
+    getName() {
+        return 'correct-twig-component-structure';
+    }
+
+    test(data) {
+        parseOutputFieldHelper(data.components).forEach(component => {
+            const { files, type, name, path } = component;
+
+            if (files.twig.exists) {
```
Comment (FRITS-Kh) [inline rules/correct-twig-component-structure.js]: Also please use early return.

File: rules/correct-twig-component-structure.js [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+                let configName;
+                let configNameStringIndex;
+                let isAtomicDesignEntityExtention;
+                let isModelComponentExtention;
+
+                if (configDefinition) {
+                    const { contract } = configDefinition;
+                    configNameStringIndex = contract.indexOf('name:');
+
+                    if (configNameStringIndex >= 0) {
```
Comment (FRITS-Kh) [inline rules/correct-twig-component-structure.js]: Suggest to move to a higher level with condition `configDefinition && configNameStringIndex >= 0`.

File: rules/correct-twig-component-structure.js [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+
+                if (extendsIndex) {
+                    this.outcome.addError(`Twig file should begin with extend in ${component.type} ${bold(component.name)}:\n${dim(component.path)}`);
+                }
+
+                if (extendStrings && !isAtomicDesignEntityExtention && !isModelComponentExtention) {
+                    this.outcome.addError(`Component template should extend atom, molecule, organism or model component in ${component.type} ${bold(component.name)}:\n${dim(component.path)}`);
+                }
+
+                if (extendStrings && isModelComponentExtention && blocks.length && !blocks.filter(block => block.name === 'body').length) {
```
Comment (FRITS-Kh) [inline rules/correct-twig-component-structure.js]: Suggest to create a variable with clear name for this condition.

File: rules/correct-typescript-component-structure.js [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+    getName() {
+        return 'correct-typescript-component-structure';
+    }
+
+    test(data) {
+        parseOutputFieldHelper(data.components).forEach(component => {
+            const { files, type, name, path } = component;
+
+            if (files.typescript.exists) {
+                if (!component.files.typescript.api.external) {
```
Comment (FRITS-Kh) [inline rules/correct-typescript-component-structure.js]: Please combine in the one condition for early return.

**Review by FRITS-Kh — CHANGES_REQUESTED** [review]

File: rules/correct-twig-component-structure.js (hunk 16 above)
Comment (supproduction) [reply rules/correct-twig-component-structure.js]: Removed unnecessary inspection variable

**Review by FRITS-Kh — APPROVED** [review]
